### ISSUE TITLE
Server is busy looping on shutdown.

### ISSUE THREAD

**Issue by alexvy86** [grpc/grpc#13958]
When letting a GRPC server shut down cleanly, I'm seeing a barrage of errors when `GRPC_TRACE` includes `op_failure`.
 
### What version of gRPC and what language are you using?
1.8.3, C# 
 
### What operating system (Linux, Windows, …) and version?
Windows 10 Pro 1709, and also happens on Docker containers on Ubuntu 16.04.2 LTS.
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
Using the C# libraries.
 
### What did you do?
Simplest repro: ran the Greeter sample application with `GRPC_TRACE` set to `op_failure` and pressed Enter to stop the GreeterServer. 
 
### What did you expect to see?
~~No errors. I just started the server, established no connections to it, and let it shutdown.~~
Per @nicolasnoble's comments below, seeing a couple of errors in the trace is probably expected, as long as they don't surface in the API. But not 3000+ copies of the same error, like we're seeing right now.
 
### What did you see instead?
A flood (3000+) of errors, all the same message:
`E0109 12:50:21.757118 0 C:\jenkins\workspace\gRPC_build_artifacts\platform\windows\workspace_csharp_ext_windows_x64\src\core\lib\surface\completion_queue.cc:659: Operation failed: tag=000001DD27F487A0, error={"created":"@1515523821.047000000","description":"Server Shutdown","file":"C:\jenkins\workspace\gRPC_build_artifacts\platform\windows\workspace_csharp_ext_windows_x64\src\core\lib\surface\server.cc","file_line":1249}`
 
### Anything else we should know about your project / environment?
As stated above, I'm seeing this when running the app directly on a Windows host, or inside Linux docker containers. ~~[Here](https://github.com/alexvy86/grpc/tree/infinite-server-errors)'s a fork of the sample app with GRPC_TRACE set. Just run the GreeterServer app from Visual Studio and press ENTER.~~ broken after rebase on top of upstream's latest changes, new link in the most recent comment.

**Comment by alexvy86** [comment]
Still seeing this error after merging upstream changes from this repo, and upgrading the NuGet dependencies as follows:
- `Google.Protobuf` to version 3.5.1
- `Google.Protobuf.Tools` to version 3.5.1
- `Grpc` to version 1.10..1
- `Grpc.Tools` to version 1.10.1

Slight changes in the error message:
`E0408 22:32:41.514949 0 T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\lib\surface\completion_queue.cc:634: Operation failed: tag=000002012FEA3030, error={"created":"@1523244758.400000000","description":"Server Shutdown","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\lib\surface\server.cc","file_line":1208}`

[Updated reference](https://github.com/alexvy86/grpc/tree/9b7a53aee22aff87d1d0d22e41d78c863f5eba1a/examples/csharp/helloworld-from-cli) to sample repro in my forked repo.

**Comment by ferrouswheel** [comment]
Happens with python on Win10 too, so not just confined to C#.

**Comment by nicolasnoble** [comment]
So... what is wrong exactly ? Tracing is for debugging. You're seeing the server doing things properly, and on shutdown, it's cancelling a bunch of things which will result in errors bubbling up - which would be a problem if you didn't actually requested a shutdown.

This seems to be a typical case of false positive.

**Comment by ferrouswheel** [comment]
"Error" implies undesired behavior. If it's impossible to shutdown the
server without errors, then maybe error is not the right term.

**Comment by nicolasnoble** [comment]
I'll have to disagree. These errors are emitted as they should be, and just like with any other piece of code that is able to throw and catch exceptions, the system is catching them, evaluates them in their context, and moves on accordingly. As long as you're not getting an error being reported by the surface API into your own code, everything's fine.

Compare this to the simple comparison between the Unix `rm` and `rm -f` operations. If I try to `rm` a non-existent file, I get an error that bubbles up to my shell. If however I try to `rm -f` a non-existent file, no error is being reported to my shell. However, if I enable debug tracing on both, I can still see the error cropping up:

```
$ strace -f rm -f foo
execve("/bin/rm", ["rm", "-f", "foo"], [/* 50 vars */]) = 0
[...]
newfstatat(AT_FDCWD, "foo", 0xc7c368, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
unlinkat(AT_FDCWD, "foo", 0)            = -1 ENOENT (No such file or directory)
lseek(0, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
+++ exited with 0 +++
```

Do I see the system throwing and catching errors when debug-tracing it ? Sure. Is it an indication something is actually wrong ? No.

**Comment by alexvy86** [comment]
@nicolasnoble I see your point, but is it really expected to see that many identical ones, non-stop? They don't come up every second or something like that, they immediately flood the screen. I just counted 3000 by copying them to Notepad++, and my console buffer actually maxed out just on that, so there could be many more.

**Comment by nicolasnoble** [comment]
Okay, that's more interesting then. Let me adjust your bug report somewhat.

**Comment by sreecha** [comment]
Sorry for the delay in responding. This is expected behavior.

**TLDR version**
- Grpc Server calls the grpc-core API `grpc_server_request_call` a few thousand times when it starts to get the ball rolling. (That API is expected to be called for each expected incoming call)
- When the server shutdowns, all those pending `grpc_server_request_call` requests "fail" showing that error. 
It is perfectly fine

**Detailed version**
Here are some relevant details:
- C# server (or any server that is wrapped around grpc-core library) must call  the grpc-core api [`grpc_server_request_call` ](https://github.com/grpc/grpc/blob/v1.14.0-pre1/include/grpc/grpc.h#L333) for *every* incoming call the Server expects. 
- Note that for every `grpc_server_request_call`, the grpc-core library will allow one incoming request and deliver it to the C# Server.  As such, the recommended practice for the Servers (that use grpc-core library) is to do the following:
       - Call `grpc_server_request_call` a few hundred times at startup to get the ball rolling
       - Call `grpc_server_request_call` everytime grpc-core library delivers it an incoming call (so that grpc-core library continues to deliver future incoming calls too.)

In  case of C#, the server seems to request 2000 incoming calls to begin with: (see  https://github.com/grpc/grpc/blob/v1.14.0-pre1/src/csharp/Grpc.Core/Server.cs#L36 and https://github.com/grpc/grpc/blob/v1.14.0-pre1/src/csharp/Grpc.Core/Server.cs#L157)

Anyway, now when the server is shutdown, all the outstanding `grpc_server_request_call` calls (i.e which are not matched with any actual incoming calls) have to be 'cancelled' and that is what results in that error message you are seeing.

We could try to be a bit smart in grpc-core and suppress those error messages but it is not worth it - considering that clean server shutdown's are rare (moreover, one could argue that this is expected behavior because your op i.e `grpc_server_request_call` technically failed as far as the core is concerned). Moreover this only happens when you have `op_failure` tracer enabled.

**Comment by alexvy86** [comment]
That's an... interesting, API design. @sreecha do you have any insight into why the API works that way? I'm curious why `grpc_server_request_call` was designed like that instead of just having the server register once and receive all the calls afterwards... If there's a good reason behind it, I guess it would be a good learning experience :).

**Comment by sreecha** [comment]
@alexvy86,  The exact rationale behind it predates me but apart from the benefits like extra flexibility the api provides (i.e choosing which completion queues the incoming request should be notified on, and which completion queue the call is bound to),  this provides a natural way to throttle incoming requests**  

Also, the grpc-core library strictly follows the "one tag in - one tag out"  policy (i.e the number of tags passed in to the grpc-core , match the number of tags returned by the `grpc_completion_queue_next()` API)

(** Currently, there isn't any throttling implemented at grpc-core layer - i.e if there is an incoming request but there is no earlier `grpc_server_request_call` , the incoming request is queued in grpc-core until `grpc_server_request_call` is made...but this can be very easily changed in future).

**Comment by nicolasnoble** [comment]
Here's a more detailed rationale.

If you follow the typical methodology of "get an incoming call -> allocate memory -> respond to the call -> free the memory", you have allocation in the critical path.

But if you pre allocate things, you can change the sequence to "get a call -> respond to the call -> free memory -> allocate memory".

This way, you moved the allocation to the tail end of your call sequence, away from the critical path, thus reducing the time between the moment you receive a call and the moment you respond to it.

Overall it's the same amount of work, but the critical path is uncluttered, and all of the clutter is moved to the tail.

**Comment by alexvy86** [comment]
Getting allocation out of the critical path makes sense... although it still feels like that could be done without the need for "outstanding call requests", or needing the caller to know/estimate how many calls it expects. Anyway, I know this is not the place to question the API of something written by Google to handle their workloads, and the original question was already answered so I'll leave it at that. Thanks to both of you for the explanations!
